webrtc/libjingle/xmpp/boshsocket.cc

```diff
@@ -0,0 +1,602 @@
+    rtc::Thread* pth = rtc::Thread::Current();
+    primary_socket_ = new ManagedSocket();
+    secondary_socket_ = new ManagedSocket();
+    if (family == AF_UNSPEC) {
+      family = AF_INET;
+    }
+    rtc::AsyncSocket* first_socket =
+      pth->socketserver()->CreateAsyncSocket(family, SOCK_STREAM);
+    rtc::AsyncSocket* second_socket = 
+      pth->socketserver()->CreateAsyncSocket(family, SOCK_STREAM);
+#ifndef USE_SSLSTREAM
+#ifdef FEATURE_ENABLE_SSL
+    if (tls_ != buzz::TLS_DISABLED) {
+      first_socket = rtc::SSLAdapter::Create(first_socket);
+      second_socket = rtc::SSLAdapter::Create(second_socket);
+    }
+#endif  // FEATURE_ENABLE_SSL
+    primary_socket_->socket_ = first_socket;
+    secondary_socket_->socket_ = second_socket;
+
+    primary_socket_->socket_->SignalReadEvent.connect(this, &BoshSocket::OnReadEvent);
+    primary_socket_->socket_->SignalWriteEvent.connect(this, &BoshSocket::OnWriteEvent);
+    primary_socket_->socket_->SignalConnectEvent.connect(this,
+      &BoshSocket::OnConnectEvent);
+    primary_socket_->socket_->SignalCloseEvent.connect(this, &BoshSocket::OnCloseEvent);
+
+    secondary_socket_->socket_->SignalReadEvent.connect(this, &BoshSocket::OnReadEvent);
+    secondary_socket_->socket_->SignalWriteEvent.connect(this, &BoshSocket::OnWriteEvent);
+    secondary_socket_->socket_->SignalConnectEvent.connect(this,
+      &BoshSocket::OnConnectEvent);
+    secondary_socket_->socket_->SignalCloseEvent.connect(this, &BoshSocket::OnCloseEvent);
+
+#else  // USE_SSLSTREAM
+    primary_socket_->socket_ = first_socket;
+    secondary_socket_->socket_ = second_socket;
+    primary_socket_->stream_ = new rtc::SocketStream(primary_socket_->socket_);
+    secondary_socket_->stream_ = new rtc::SocketStream(secondary_socket_->socket_);
+#ifdef FEATURE_ENABLE_SSL
+    if (tls_ != buzz::TLS_DISABLED) {
+      primary_socket_->stream_ = rtc::SSLStreamAdapter::Create(primary_socket_->stream_);
+      secondary_socket_->stream_ = rtc::SSLStreamAdapter::Create(secondary_socket_->stream_);
+    }
+#endif  // FEATURE_ENABLE_SSL
+    primary_socket_->stream_->SignalEvent.connect(this, &BoshSocket::OnEvent);
+    secondary_socket_->stream_->SignalEvent.connect(this, &BoshSocket::OnEvent);
+#endif  // USE_SSLSTREAM
+  }
+
+
+  BoshSocket::~BoshSocket()
+  {
+    //******************Implement Close
+    Close();
+    delete generator_;
+#ifdef USE_SSLSTREAM
+    delete primary_socket_->stream_;
+    delete secondary_socket_->stream_;
+#else  // USE_SSLSTREAM
+    delete primary_socket_->socket_;
+    delete secondary_socket_->socket_;
+#endif
+    delete primary_socket_;
+    delete secondary_socket_;
+  }
+
+#ifndef USE_SSLSTREAM
+  void BoshSocket::OnReadEvent(rtc::AsyncSocket *socket) {
+    auto temp_socket = primary_socket_;
+    if (temp_socket->socket_ != socket) {
+      temp_socket = secondary_socket_;
+    }
+    last_receive_ = time(NULL);
+    buzz::InputMsg input_msg;
+    char bytes[4096];
```

Comment:
@dscherba I guess that you might point to a wrong loc. May be this line is the place that you found the problem, since the below code that you point will only appear when you set the USE_SSLSTREAM flag.
I ran many time but I never found the problem that you're facing. Can you tell me the step to reproduce it or send me some information about the problem?